## PAIR PROGRAMMING EDA II

### Nulos y Valores extremos
---

In [39]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import sidetable

#### **Hipótesis**

Antes de seguir, recordamos las preguntas que nos planteamos al principio del pair programming de EDA para dirigir nuestro análisis.

- ¿Es Australia el sitio más peligroso y letal para estar relajada en la playa?
- ¿Hay diferencias entre los paises en los ataques a hombres y mujeres?
- ¿Cuáles son las edades que más sufren ataques?
- Independientemente de la edad, ¿sufren los hombres más ataques que las mujeres?
- ¿En qué mes ocurren más ataques?
- ¿Cuál es la relación entre la especie y el tipo de ataque (si es fatal o no)?
- ¿Cómo han evolucionado los ataques a lo largo del tiempo?

Columnas a considerar:
¿Está limpia?
- age   ❌     esta en formato string cuando debería ser integer, y en algunos casos tenemos rangos de edad
- species  ❌  es un jaleo! Debemos unificar los nombres y reducir a las especies más importantes
- country ❌   los paises están en mayúsculas, algunos se repiten con algunos cambios
- fatal ✔️      la limpiamos en el pair de Pandas V
- year ❌      es una columna de tipo float, deberíamos convertirla a integer
- sex ✔️        la limpiamos en el pair de Pandas V
- fecha ✔️      la limpiamos en el pair de Pandas V

#### **Conclusiones**

¿Qué tendréis que hacer?

- En relación a los valores nulos:

        Identificar las columnas que tienen nulos (lo hicimos en el ejercicio de pair programming EDA I y en el de hoy con el sidetable).

        Debatid entre vosotras qué haríais con los valores nulos, ¿los quitaríais o los reemplazaríais con un valor que nosotras decidiremos?. Cread una celda de markdown en vuestro jupyter y escribid vuestras conclusiones.

- En relación a los outliers, ya hemos ido conociendo un poco nuestro dataframe y de momento solo tenemos una columna de tipo numérico, por lo tanto, poco podemos hacer. Un poco más adelante, cuando hayamos hecho una serie de cambios sobre nuestras columnas podremos hacer una evaluación más profunda de estos.

In [40]:
df= pd.read_pickle("../files/attacks8.pickle")

#### **Objetivos**

Usad la librería sidetable para:

In [41]:
df.dtypes

case_number    object
year            int64
type           object
country        object
activity       object
age            object
species_       object
date           object
mes_ataque     object
fatal          object
sexo           object
dtype: object

In [42]:
df.head()

case_number  year        type     country  \
0    1800.00.00  1970  Unprovoked  seychelles   
1  1797.05.28.R  1970  Unprovoked         NaN   
2    1792.09.12  1970    Provoked     england   
3    1791.00.00  1970  Unprovoked   australia   
4    1788.05.10  1970        Boat   australia   

                          activity  age species_                  date  \
0  a corsair's boat was overturned  NaN      NaN                  1800   
1                Dropped overboard  NaN      NaN  Reported May-28-1797   
2                          Fishing  NaN      NaN           12-Sep-1792   
3                              NaN  NaN      NaN                  1791   
4                          Fishing  NaN      NaN           10-May-1788   

  mes_ataque fatal sexo  
0        NaN     Y    F  
1       May      Y  NaN  
2       Sep      Y    M  
3        NaN     Y    F  
4       May      N  NaN

**EJERCICIO 1**

Explorar los valores nulos.


In [43]:
df.stb.missing().round(3)

missing  total  percent
age              154   1672    9.211
mes_ataque       141   1672    8.433
species_         126   1672    7.536
fatal             97   1672    5.801
activity          31   1672    1.854
sexo              14   1672    0.837
country           10   1672    0.598
case_number        0   1672    0.000
year               0   1672    0.000
type               0   1672    0.000
date               0   1672    0.000

    La columna con mas valores nulos es age, el tratamiento que le hariamos seria a priori mantenerlos  ya que no llegan al 10% y nos afectaria los registros de las otras columnas. Si mas adelante necesitamos ampliar el analisis con esta columna, podriamos ver de cambiar los mismos por  la mediana o la media siempre previo analisis de los outliers.

**EJERCICIO 2**

Explorad la columna de country y reportad los resultados solo para aquellas categorías que incluyan el 70% de los datos.


In [44]:
df.stb.freq(['country'], thresh= 70, other_label= 'otros paises').round(3)

country  count  percent  cumulative_count  cumulative_percent
0           usa    741   44.585               741              44.585
1     australia    343   20.638              1084              65.223
2  otros paises    578   34.777              1662             100.000

    Vemos que los paises donde hay mas registros de ataques son USA y Australia, los cuales suman el 65% de todo el conjunto de datos.

**EJERCICIO 4**

Explorad la columna country en relación al sexo, filtrando de nuevo por aquellas categorías que incluyan el 70% de los datos.


In [45]:
df.stb.freq(['sexo', 'country'], thresh= 70, other_label= 'otros paises').round(3)

sexo       country  count  percent  cumulative_count  \
0             M           usa    609   36.909               609   
1             M     australia    309   18.727               918   
2             M  south africa    224   13.576              1142   
3  otros paises  otros paises    508   30.788              1650   

   cumulative_percent  
0              36.909  
1              55.636  
2              69.212  
3             100.000

    Al filtrar el 70% de los casos vemos que nos devuelve en country nuevamente Usa, Australia y agrega South Africa y entendemos que el 70% de  los ataques se dan en hombres por eso solo nos trae M en la columna 'sexo'.

**EJERCICIO 5**

Haced los mismo para la columna activity en relación a la columna fatal. En este caso filtrad y mostrad los datos que incluyan el 60% de los datos.


In [46]:
df.stb.freq(['activity', 'fatal'], thresh= 60, other_label= 'otras actividades').round(3)

activity              fatal  count  percent  cumulative_count  \
0             Surfing                  N    380   24.500               380   
1            Swimming                  N    140    9.026               520   
2        Spearfishing                  N     85    5.480               605   
3            Swimming                  Y     79    5.093               684   
4             Fishing                  N     50    3.224               734   
5              Wading                  N     43    2.772               777   
6            Standing                  N     30    1.934               807   
7          Snorkeling                  N     29    1.870               836   
8             Surfing                  Y     26    1.676               862   
9       Body boarding                  N     24    1.547               886   
10             Diving                  N     19    1.225               905   
11       Scuba diving                  N     17    1.096               922   
12  otras actividades  otras actividades    629   40.554              1551   

    cumulative_percent  
0               24.500  
1               33.527  
2               39.007  
3               44.101  
4               47.324  
5               50.097  
6               52.031  
7               53.901  
8               55.577  
9               57.124  
10              58.349  
11              59.446  
12             100.000

    Las tres primeras actividades (Surfing, Swimming y Spearfishing) que mas cantidad de ataques tienen sin ser fatales  acumulan el 39% de los registros, volviendo a aparecer a continuacion con porcentajes menores en los casos que si son fatales.